### PR TITLE
Send buttons and Lists

### PR DESCRIPTION
**Pull request by tuyuribr**
Add list and buttons.
Thanks to @PurpShell

### CODE REVIEW

**Comment by PurpShell** [comment]
all in for that credit lol

good job

**Comment by creuza** [comment]
how can i use this new feature?

**Comment by PurpShell** [comment]
> how can i use this new feature?

it is a draft.

**Comment by tuyuribr** [comment]
Button example:
```javascript

//text button
let button = new Buttons("body",[{body:"bt1"},{body:"bt2"},{body:"bt3"}],"title","footer");
client.sendMessage(message['from'], button);


//media button ( Testing, images works fine )

let media = MessageMedia.fromFilePath("image.png");
let button = new Buttons(media,[{body:"bt1"},{body:"bt2"},{body:"bt3"}],"title","footer");
client.sendMessage(message['from'], button,{'caption':"body"});

```

**Comment by tuyuribr** [comment]
Button tests needed :
Android Business
IOS Standart and Business

**Comment by creuza** [comment]
`(node:6580) UnhandledPromiseRejectionWarning: ReferenceError: Buttons is not defined`

I received this, even updating the project

**Comment by PurpShell** [comment]
> `(node:6580) UnhandledPromiseRejectionWarning: ReferenceError: Buttons is not defined`
> 
> I received this, even updating the project

have you cloned the PR????? are you using NPM?

**Comment by creuza** [comment]
i use npm

**Comment by creuza** [comment]
now it worked!

`const { Client,  Buttons } = require('whatsapp-web.js');`

it was missing to declare the Buttons

**Comment by tuyuribr** [comment]
List example:

```javascript
let sections = [{title:'sectionTitle',rows:[{id:'customId', title:'ListItem1', description: 'desc'},{title:'ListItem2'}]}];
let list = new List('aaa','btnText',sections,'Title','footer');
client.sendMessage(message['from'], list);
```
Lists aren't working on business right now

Tests needed:
Ios Standard

**Comment by esteves67** [comment]
> Button example:
> 
> ```js
> //text button
> let button = new Buttons("body",[{body:"bt1"},{body:"bt2"},{body:"bt3"}],"title","footer");
> client.sendMessage(message['from'], button);
> 
> 
> //media button ( Testing, images works fine )
> 
> let media = MessageMedia.fromFilePath("image.png");
> let button = new Buttons(media,[{body:"bt1"},{body:"bt2"},{body:"bt3"}],"title","footer");
> client.sendMessage(message['from'], button,{'caption':"body"});
> ```
> 
> Button tests needed :
> Android Business
> IOS Standard and Business

Tested on Android Business, and iOS Standard, and its working.
Idk about iOS Business.

**Comment by esteves67** [comment]
I've been testing this PR, but faced into a problem, i know this PR isn't ready yet, but just wanted you guys to know.

When you have a conversation with buttons, if you use client.getChats you get a problem with map (https://github.com/pedroslopez/whatsapp-web.js/issues/297)

**Comment by esteves67** [comment]
I've been testing this DRAFT, and took notes of some things.

Number 1:
When you have a conversation with buttons (using this PR), if you use client.getChats you get a problem with map (https://github.com/pedroslopez/whatsapp-web.js/issues/297)
The only solution for now its to delete the whole conversation that has buttons in it, and the map function works again.
If you guys can test it, test it out, just to make sure its not my fault in my client.getChats function.

Number 2:
Buttons are not getting logged on the console, only the main message gets logged to the console, i guess its not being recognized by whatsapp-web.js yet.

**Comment by PurpShell** [comment]
> I've been testing this DRAFT, and took notes of some things.
> 
> Number 1:
> When you have a conversation with buttons (using this PR), if you use client.getChats you get a problem with map (#297)
> The only solution for now its to delete the whole conversation that has buttons in it, and the map function works again.
> If you guys can test it, test it out, just to make sure its not my fault in my client.getChats function.
> 
> Number 2:
> Buttons are not getting logged on the console, only the main message gets logged to the console, i guess its not being recognized by whatsapp-web.js yet.

Makes more sense since buttons_message is way different than templateMessage which msg.buttons is trying to act with

**Comment by tuyuribr** [comment]
> 
> 
> I've been testing this DRAFT, and took notes of some things.
> 
> Number 1:
> When you have a conversation with buttons (using this PR), if you use client.getChats you get a problem with map (#297)
> The only solution for now its to delete the whole conversation that has buttons in it, and the map function works again.
> If you guys can test it, test it out, just to make sure its not my fault in my client.getChats function.
> 
> Number 2:
> Buttons are not getting logged on the console, only the main message gets logged to the console, i guess its not being recognized by whatsapp-web.js yet.

Thanks for your help!

**Comment by esteves67** [comment]
> I've been testing this DRAFT, and took notes of some things.
> 
> Number 1:
> When you have a conversation with buttons (using this PR), if you use client.getChats you get a problem with map (#297)
> The only solution for now its to delete the whole conversation that has buttons in it, and the map function works again.
> If you guys can test it, test it out, just to make sure its not my fault in my client.getChats function.
> 
> Number 2:
> Buttons are not getting logged on the console, only the main message gets logged to the console, i guess its not being recognized by whatsapp-web.js yet.

Ignore number 1, as it is already solved by https://github.com/pedroslopez/whatsapp-web.js/issues/386

**Comment by iamshouvikmitra** [comment]
Any idea when we can have this shipped?
Super excited!

File: src/structures/Buttons.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        this.title = title;
+        
+        /**
+         * footer of message
+         * @type {string}
+         */
+        this.footer = footer;
+
+        if (body instanceof MessageMedia) {
+            // throw '[BT02] MessageMedia not ready';
```
Comment (PurpShell) [inline src/structures/Buttons.js:36]: kill this.

**Review by purpshell — CHANGES_REQUESTED** [review]

**Review by purpshell — APPROVED** [review]
LGTM!

**Comment by 0xT01** [comment]
send list/buttons error
```
(node:979379) UnhandledPromiseRejectionWarning: TypeError: Util.generateHash is not a function
at rows.section.rows.map (whatsapp-web.js/src/structures/List.js:68:55)
```

how to fix.?

**Comment by PurpShell** [comment]
> send list/buttons error
> 
> ```
> (node:979379) UnhandledPromiseRejectionWarning: TypeError: Util.generateHash is not a function
> at rows.section.rows.map (whatsapp-web.js/src/structures/List.js:68:55)
> ```
> 
> how to fix.?

the util functions have disappeared on your end? This is not an issue, since Util is required on the top of the file.

**Comment by ricocrivelli** [comment]
How can I help to test this functionality?

**Comment by 0xT01** [comment]
Hey guys.
I've tried send message list and button,
for a smooth list to be sent.
I tried to send a button message but it didn't send even there was no error message from the system, it looks normal but the message is not sent, I use personal whatsapp.

responses:
```
Message {
  mediaKey: undefined,
  id:
   { fromMe: true,
     remote:
      { server: 'c.us',
        user: '62xxxx',
        _serialized: '62xxxx@c.us' },
     id: '3EB0CB3BE1BB703750DD',
     _serialized: 'true_628xxxx@c.us_3EB0CB3BE1BB703750DD' },
  ack: 0,
  hasMedia: false,
  body: 'ini test body button',
  type: 'chat',
  timestamp: 1628296215,
  from: '62xxxx@c.us',
  to: '62xxxx@c.us',
  author: undefined,
  deviceType: 'web',
  isForwarded: false,
  forwardingScore: 0,
  isStatus: undefined,
  isStarred: false,
  broadcast: undefined,
  fromMe: true,
  hasQuotedMsg: false,
  location: undefined,
  vCards: [],
  inviteV4: undefined,
  mentionedIds: [],
  orderId: undefined,
  token: undefined,
  title: 'ini test title',
  links: undefined,
  dynamicReplyButtons:
   [ { buttonId: 'ySE9Te', buttonText: [Object], type: 1 },
     { buttonId: 'VkcgC7', buttonText: [Object], type: 1 },
     { buttonId: 'JiwHP8', buttonText: [Object], type: 1 } ] }
```
please give directions so that the message button can be sent. thank you

**Comment by tonvital** [comment]
@0xT01 
i think only whatsapp business can send buttons and lists.

**Comment by 0xT01** [comment]
@tonvital 
I've tried using WhatsApp for business. and the result list message is not sent. only personal whatsapp can and the list message is sent. this is a button message that has not got a solution

**Comment by 0xT01** [comment]
Screenshot_2021-08-07-21-21-27-07_7352322957d4404136654ef4adb64504
Screenshot_2021-08-07-21-21-30-95_7352322957d4404136654ef4adb64504

**Comment by tuyuribr** [comment]
> 
> 
> @0xT01
> i think only whatsapp business can send buttons and lists.

Nope, WaBiz can't send lists because something is off.
I'm sending buttons rn from a standard account

image

**Comment by 0xT01** [comment]
@tuyuribr

yes. but the button doesn't work on the standard account I have.

[0xT01 ](https://github.com/pedroslopez/whatsapp-web.js/pull/756#issuecomment-894578698l)

> Hey guys.
> I've tried send message list and button,
> for a smooth list to be sent.
> I tried to send a button message but it didn't send even there was no error message from the system, it looks normal but the message is not sent, I use personal whatsapp.
> 
> responses:
> 
> ```
> Message {
>   mediaKey: undefined,
>   id:
>    { fromMe: true,
>      remote:
>       { server: 'c.us',
>         user: '62xxxx',
>         _serialized: '62xxxx@c.us' },
>      id: '3EB0CB3BE1BB703750DD',
>      _serialized: 'true_628xxxx@c.us_3EB0CB3BE1BB703750DD' },
>   ack: 0,
>   hasMedia: false,
>   body: 'ini test body button',
>   type: 'chat',
>   timestamp: 1628296215,
>   from: '62xxxx@c.us',
>   to: '62xxxx@c.us',
>   author: undefined,
>   deviceType: 'web',
>   isForwarded: false,
>   forwardingScore: 0,
>   isStatus: undefined,
>   isStarred: false,
>   broadcast: undefined,
>   fromMe: true,
>   hasQuotedMsg: false,
>   location: undefined,
>   vCards: [],
>   inviteV4: undefined,
>   mentionedIds: [],
>   orderId: undefined,
>   token: undefined,
>   title: 'ini test title',
>   links: undefined,
>   dynamicReplyButtons:
>    [ { buttonId: 'ySE9Te', buttonText: [Object], type: 1 },
>      { buttonId: 'VkcgC7', buttonText: [Object], type: 1 },
>      { buttonId: 'JiwHP8', buttonText: [Object], type: 1 } ] }
> ```
> 
> please give directions so that the message button can be sent. thank you

**Comment by tuyuribr** [comment]
> 
> 
> @tuyuribr
> 
> yes. but the button doesn't work on the standard account I have.
@0xT01 
Try again but with the new commit ( c3becb8 ), get the pr again

**Review by Gimenz — APPROVED** [review]

**Comment by ricocrivelli** [comment]
> 
> 
> > @tuyuribr
> > yes. but the button doesn't work on the standard account I have.
> > @0xT01
> > Try again but with the new commit ( [c3becb8](https://github.com/pedroslopez/whatsapp-web.js/commit/c3becb8ec0940bdfce50f9abb3d6b89911d8504d) ), get the pr again

Its working now!!

**Comment by 0xT01** [comment]
**WhatsApp** has a new feature. only displays once for the image.?
can this feature be added here?

WhatsApp Image 2021-08-18 at 01 18 47
WhatsApp Image 2021-08-18 at 01 18 48
WhatsApp Image 2021-08-18 at 01 18 49
WhatsApp Image 2021-08-18 at 01 18 49 (1)
WhatsApp Image 2021-08-18 at 01 18 50

**Comment by tuyuribr** [comment]
> 
> 
> **WhatsApp** has a new feature. only displays once for the image.?
> can this feature be added here?
> 

In the lib: Yes
in this PR: No
Hotel? Triv@g0

**Comment by Spooder9** [comment]
Hi, it doesn't work for me, what am I doing wrong?
Captura de pantalla de 2021-08-20 09-20-46

**Comment by ricocrivelli** [comment]
> 
> 
> Hi, it doesn't work for me, what am I doing wrong?
> Captura de pantalla de 2021-08-20 09-20-46

You probably forgot to import the class at the beginning of the code.

**Comment by Spooder9** [comment]
> > Hi, it doesn't work for me, what am I doing wrong?
> > Captura de pantalla de 2021-08-20 09-20-46
> 
> You probably forgot to import the class at the beginning of the code.
> 
> `const { ClientInfo, Message, MessageMedia, Contact, Location, GroupNotification , Label, Call, Buttons, List} = require('./structures');`

I have already imported the module as you indicate but it still does not work.
Captura de pantalla de 2021-08-20 11-00-25

**Comment by Spooder9** [comment]
Captura de pantalla de 2021-08-20 11-03-10

**Comment by ricocrivelli** [comment]
> 
> 
> Captura de pantalla de 2021-08-20 11-03-10

Looks like you are using the production source. This is a pull request, so the new source isn't in the main files. You can clone the branch https://github.com/tuyuribr/whatsapp-web.js/tree/patch-6 ou see the modified files and create by your own https://github.com/pedroslopez/whatsapp-web.js/pull/756/files

**Comment by Scoooolz** [comment]
How i can fix this?

(node:23484) UnhandledPromiseRejectionWarning: TypeError: Buttons is not a constructor
    at Client.<anonymous> (C:\Users\User\whatsapp-api\app.js:70:26)
    at Client.emit (events.js:375:28)
    at C:\Users\User\whatsapp-api\node_modules\whatsapp-web.js\src\Client.js:251:18
    at Page._onBindingCalled (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:980:62)
    at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:159:60
    at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
    at Array.map (<anonymous>)
    at Object.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
    at CDPSession.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)
    at CDPSession._onMessage (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:234:18)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:23484) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:23484) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by ricocrivelli** [comment]
> 
> 
> How i can fix this?
> 
> (node:23484) UnhandledPromiseRejectionWarning: TypeError: Buttons is not a constructor
> at Client. (C:\Users\User\whatsapp-api\app.js:70:26)
> at Client.emit (events.js:375:28)
> at C:\Users\User\whatsapp-api\node_modules\whatsapp-web.js\src\Client.js:251:18
> at Page._onBindingCalled (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:980:62)
> at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:159:60
> at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
> at Array.map ()
> at Object.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
> at CDPSession.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)
> at CDPSession._onMessage (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:234:18)
> (Use `node --trace-warnings ...` to show where the warning was created)
> (node:23484) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:23484) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

Take a look at my previous comment, maybe it can help you.  https://github.com/pedroslopez/whatsapp-web.js/pull/756#issuecomment-902965071

**Comment by Scoooolz** [comment]
> > How i can fix this?
> > (node:23484) UnhandledPromiseRejectionWarning: TypeError: Buttons is not a constructor
> > at Client. (C:\Users\User\whatsapp-api\app.js:70:26)
> > at Client.emit (events.js:375:28)
> > at C:\Users\User\whatsapp-api\node_modules\whatsapp-web.js\src\Client.js:251:18
> > at Page._onBindingCalled (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:980:62)
> > at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:159:60
> > at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
> > at Array.map ()
> > at Object.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
> > at CDPSession.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)
> > at CDPSession._onMessage (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:234:18)
> > (Use `node --trace-warnings ...` to show where the warning was created)
> > (node:23484) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> > (node:23484) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> 
> Take a look at my previous comment, maybe it can help you. [#756 (comment)](https://github.com/pedroslopez/whatsapp-web.js/pull/756#issuecomment-902965071)

Same.

(node:30828) UnhandledPromiseRejectionWarning: TypeError: Buttons is not a constructor
    at Client.<anonymous> (C:\Users\User\whatsapp-api\app.js:73:26)
    at Client.emit (events.js:375:28)
    at C:\Users\User\whatsapp-api\node_modules\whatsapp-web.js\src\Client.js:251:18
    at Page._onBindingCalled (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:980:62)
    at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:159:60
    at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
    at Array.map (<anonymous>)
    at Object.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
    at CDPSession.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)
    at CDPSession._onMessage (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:234:18)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:30828) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:30828) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

Are i wrong?

In const
const { Client } = require('whatsapp-web.js');
const { Buttons } = require('./node_modules/whatsapp-web.js/src/structures')
const express = require('express');
const io = require('socket.io')
const qrcode = require('qrcode-terminal');
const fs = require('fs');

In Commands
client.on('message', async msg => {

        if(msg.body === '!quiz') {
            let button = new Buttons('Button body',[{body:'bt1'},{body:'bt2'},{body:'bt3'}],'title','footer');
            client.sendMessage(msg.from, button);
});

**Comment by Scoooolz** [comment]
> Take a look at my previous comment, maybe it can help you. [#756 (comment)](https://github.com/pedroslopez/whatsapp-web.js/pull/756#issuecomment-902965071)

Same.

(node:30828) UnhandledPromiseRejectionWarning: TypeError: Buttons is not a constructor
    at Client.<anonymous> (C:\Users\User\whatsapp-api\app.js:73:26)
    at Client.emit (events.js:375:28)
    at C:\Users\User\whatsapp-api\node_modules\whatsapp-web.js\src\Client.js:251:18
    at Page._onBindingCalled (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:980:62)
    at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:159:60
    at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
    at Array.map (<anonymous>)
    at Object.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
    at CDPSession.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)
    at CDPSession._onMessage (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:234:18)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:30828) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:30828) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

Are i wrong?

In const

const { Client } = require('whatsapp-web.js');
const { Buttons } = require('./node_modules/whatsapp-web.js/src/structures')
const express = require('express');
const io = require('socket.io')
const qrcode = require('qrcode-terminal');
const fs = require('fs');


In Commands

client.on('message',`` async msg => {
        if(msg.body === '!quiz') {
            let button = new Buttons('Button body',[{body:'bt1'},{body:'bt2'},{body:'bt3'}],'title','footer');
            client.sendMessage(msg.from, button);
});

**Comment by PurpShell** [comment]
> How can I help to test this functionality?

to install & therefore test it: 

you have to npm install it via a different route
npm i github:tuyuribr/whatsapp-web.js#patch-6

btw this is how i got #patch-6 at the end:
image

**Comment by PurpShell** [comment]
> > Take a look at my previous comment, maybe it can help you. [#756 (comment)](https://github.com/pedroslopez/whatsapp-web.js/pull/756#issuecomment-902965071)
> 
> Same.
> 
> (node:30828) UnhandledPromiseRejectionWarning: TypeError: Buttons is not a constructor
> at Client. (C:\Users\User\whatsapp-api\app.js:73:26)
> at Client.emit (events.js:375:28)
> at C:\Users\User\whatsapp-api\node_modules\whatsapp-web.js\src\Client.js:251:18
> at Page._onBindingCalled (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:980:62)
> at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:159:60
> at C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
> at Array.map ()
> at Object.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
> at CDPSession.emit (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)
> at CDPSession._onMessage (C:\Users\User\whatsapp-api\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:234:18)
> (Use `node --trace-warnings ...` to show where the warning was created)
> (node:30828) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:30828) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> 
> Are i wrong?
> 
> In const
> 
> const { Client } = require('whatsapp-web.js');
> const { Buttons } = require('./node_modules/whatsapp-web.js/src/structures')
> const express = require('express');
> const io = require('socket.io')
> const qrcode = require('qrcode-terminal');
> const fs = require('fs');
> 
> In Commands
> 
> client.on('message',`` async msg => {
> if(msg.body === '!quiz') {
> let button = new Buttons('Button body',[{body:'bt1'},{body:'bt2'},{body:'bt3'}],'title','footer');
> client.sendMessage(msg.from, button);
> });

yeah you are: 
In const: 
```diff
+ const { Client, Buttons } = require('whatsapp-web.js');
- const { Buttons } = require('./node_modules/whatsapp-web.js/src/structures')
const express = require('express');
const io = require('socket.io')
const qrcode = require('qrcode-terminal');
const fs = require('fs');
```

**Comment by Spooder9** [comment]
> > Captura de pantalla de 2021-08-20 11-03-10
> 
> Looks like you are using the production source. This is a pull request, so the new source isn't in the main files. You can clone the branch https://github.com/tuyuribr/whatsapp-web.js/tree/patch-6 ou see the modified files and create by your own https://github.com/pedroslopez/whatsapp-web.js/pull/756/files

Gracias

**Comment by renjop** [comment]
> 
> 
> > It seems that the most recent WhatsApp version (2.2134.9) broke the sending of this buttons:
> > `{ "message": "Msg", "name": "ModelCreateError", "stack": "ModelCreateError: Msg\n at new i (https://web.whatsapp.com/bootstrap_qr.abbc2ea8366e2cd9a145.js:2:403072)\n at new k (https://web.whatsapp.com/bootstrap_qr.abbc2ea8366e2cd9a145.js:56:209667)\n at https://web.whatsapp.com/bootstrap_qr.abbc2ea8366e2cd9a145.js:2:470293\n at V (https://web.whatsapp.com/bootstrap_qr.abbc2ea8366e2cd9a145.js:2:373738)\n at MutationObserver.K (https://web.whatsapp.com/bootstrap_qr.abbc2ea8366e2cd9a145.js:2:373319)",`
> 
> Hmmm it's working here, are you using whatsapp-web.js@1.14.1 ?

Yep, I'm using the latest whatsapp-web.js@1.41.1. Before the WhatsApp update it was working perfectly, This error seems to be an internal "error" from WhatsApp not necessarily from the whatsapp-web.js library.

**Comment by makanSukros** [comment]
Not Working on MD Version @PurpShell

**Comment by makanSukros** [comment]
> Not Working on MD Version @PurpShell

vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34 TypeError: Cannot read properties of undefined (reading 'length')
    at bootstrap_main.e39859bf8358eb1e7271.js:13
    at Array.some (<anonymous>)
    at t.default (bootstrap_main.e39859bf8358eb1e7271.js:13)
    at ia (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Hs (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Pu (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Eu (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Su (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at mu (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34

**Comment by makanSukros** [comment]
> Not Working on MD Version @PurpShell

```
vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34 TypeError: Cannot read properties of undefined (reading 'length')
    at bootstrap_main.e39859bf8358eb1e7271.js:13
    at Array.some (<anonymous>)
    at t.default (bootstrap_main.e39859bf8358eb1e7271.js:13)
    at ia (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Hs (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Pu (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Eu (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at Su (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at mu (vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34)
    at vendor1~bootstrap_qr.8d5b7620b726968ba154.js:34
```

**Comment by jmas005** [comment]
Can I identify the custom ID returned from a Message of type List?

**Comment by rmaafs** [comment]
I tested it for two weeks and I don't have any problem.
Tested on Android and iOS. Nice work!

**Comment by makanSukros** [comment]
> I tested it for two weeks and I don't have any problem.
> Tested on Android and iOS. Nice work!

Problem only for Beta MultiDevices

**Review by Inevitable-Dev — APPROVED** [review]
Oh, I spent some time testing some things to see if I could come up with a solution, but I wasn't able to come up with this excellent solution

**Review by aliyss — APPROVED** [review]
lgtm

**Comment by PurpShell** [comment]
Merge time

**Comment by Fibonacci-Solucoes-Ageis** [comment]
When i sent a button message the Whatsapp Web chat throws the following exception.

image

**Comment by Goomer2** [comment]
When I send a Button Message the Whatsapp Web Chat throws  the following exception.

image

**Comment by tuyuribr** [comment]
@Goomer2 , could you send us some code?
Are you using standard wa or biz?


Image for reference:
image

**Comment by Goomer2** [comment]
> @Goomer2 , could you send us some code?
> Are you using standard wa or biz?
> 
> Image for reference:
> image

I'm using biz.

```
const { Client, Buttons, List } = require('whatsapp-web.js');
const qrcode = require('qrcode-terminal');

const client = new Client({puppeteer: {headless: false}});

client.on('qr', (qr) => {
  // Generate and scan this code with your phone
  console.log('QR RECEIVED', qr);
  qrcode.generate(qr, {small: true});
});

client.on('ready', () => {
  console.log('Client is ready!');

  let button = new Buttons('Button body', [{body:'Sim'},{body:'Não'},{body:'Não enviar mais'}],'title','footer');
  client.sendMessage('xxxxx@c.us', button);
});

client.on('message', msg => {
  if (msg.body == '!ping') {
    msg.reply('pong');
  }
});

client.initialize();
```

The error is on Whatsapp Web.

**Comment by tinyCoder32** [comment]
> When I send a Button Message the Whatsapp Web Chat throws the following exception.
> 
> image

Is there any fix for this yet? It's pretty annoying 😓

**Pull request merged**